### PR TITLE
bpo-34395: Don't free allocated memory on realloc fail in load_mark() in _pickle.c.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://bugs.python.org/issue34395

### CODE REVIEW

**Review by serhiy-storchaka — APPROVED** [review]

File: Modules/_pickle.c [hunk 1]
```diff
@@ -6297,14 +6297,13 @@ load_mark(UnpicklerObject *self)
             return -1;
         }
```
Comment (serhiy-storchaka) [inline Modules/_pickle.c]: Seems the check above can be removed. `alloc > (PY_SSIZE_T_MAX / sizeof(Py_ssize_t))` is duplicated in `PyMem_RESIZE()`, and failing `alloc <= ((size_t)self->num_marks + 1)` is not possible if `sizeof(Py_ssize_t)` > 2.

Comment (sir-sigurd) [reply Modules/_pickle.c]: https://github.com/python/cpython/blob/1590c393360df059160145e7475754427bfc6680/Modules/_pickle.c#L6289
Also, it seems that new space is needed only when `self->num_marks + 1` is strictly greater than `self->marks_size`.

Comment (serhiy-storchaka) [reply Modules/_pickle.c]: Agree. This condition can be written as `self->num_marks >= self->marks_size` or `self->num_marks == self->marks_size`.

Comment (sir-sigurd) [reply Modules/_pickle.c]: @serhiy-storchaka 
I'd prefer to make these change as separate commits/RPs as they are not related to this bpo issue.

Comment (sir-sigurd) [reply Modules/_pickle.c]: PR for upsizing issue https://github.com/python/cpython/pull/8860.

**Review by zhangyangyu — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
I don't think there is a bug that needs to be fixed. This change looks rather as a code clean up to me. If I'm wrong, and it fixes a real bug, it should be backported.

@sir-sigurd, please remove also the redundant check before PyMem_RESIZE if it is truly unneeded.

**Comment by sir-sigurd** [comment]
@serhiy-storchaka 

I'm considering adding this as explanation why overflow check is not needed
```
assert((size_t)self->num_marks <= PY_SSIZE_T_MAX / sizeof(Py_ssize_t));
Py_BUILD_ASSERT((size_t)PY_SSIZE_T_MAX / sizeof(Py_ssize_t) << 1 <= SIZE_MAX - 20);
```
but I'm not sure if it makes things clear.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by zhangyangyu** [comment]
`failing alloc <= ((size_t)self->num_marks + 1) is not possible if sizeof(Py_ssize_t) > 2.`

Can this be guaranteed in Python? Wikipedia says size_t is at least 16 bit wise.

**Comment by serhiy-storchaka** [comment]
I think we can assume that `size_t` is at least 32 bit wise, and `char` is 8 bit wise. Many things in CPython implementation will be broken if this is not true.

**Comment by zhangyangyu** [comment]
> If I'm not mistaken, it can't overflow even if sizeof(Py_ssize_t) == 2 (presuming that CPython allocates at most PY_SSIZE_T_MAX bytes).

I don't get it, how it's not possible.

And rechecking the patch it seems to me the overflow check is not right. You should check overflow before doing math operation. `size_t alloc = ((size_t)self->num_marks << 1) + 20;` might already overflowed and succeed in the overflow check in `PyMem_Resize`. What do you think about it? Am I mistaken something?

**Comment by sir-sigurd** [comment]
@zhangyangyu 
Presuming that CPython allocates at most `PY_SSIZE_T_MAX` bytes, we can assume that `(size_t)self->num_marks <= PY_SSIZE_T_MAX / sizeof(Py_ssize_t)`. So 
`alloc <= (PY_SSIZE_T_MAX / sizeof(Py_ssize_t) << 1) + 20`. If `sizeof(Py_ssize_t) == 2`, then `alloc <= PY_SSIZE_T_MAX + 20` and it's obviously less than `SIZE_MAX`.

Makes sense?

**Comment by zhangyangyu** [comment]
@sir-sigurd Good insight! Thanks for the explanation.

**Pull request merged**